File: Data Structures & Algorithms/network-delay-time/submission-9.java
class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {

        PriorityQueue<int[]> minHeap = new PriorityQueue<>(
            (a,b) -> Integer.compare(a[1],b[1])
        );

        Map<Integer,List<int[]>> adj = new HashMap<>();
        for(int i=1;i<=n;i++){
            adj.put(i,new ArrayList<>());
        }
        for(int[] time:times){
            int src= time[0],dest=time[1],cost=time[2];
            adj.get(src).add(new int[]{dest,cost});
        }
        int res = 0;
        minHeap.offer(new int[]{k,0});
        Set<Integer> vis = new HashSet<>();
        while(!minHeap.isEmpty() && vis.size()<n){
            int[] time = minHeap.poll();
            int src= time[0],cost = time[1];
            if(vis.contains(src))
                continue;
            res = cost;
            vis.add(src);
            for(int[] nextTime:adj.get(src)){
                int dest = nextTime[0], destTime = nextTime[1];
                minHeap.offer(new int[]{dest,cost+destTime});
            }
        }
        return vis.size()==n?res:-1;


        
    }
}

File: Data Structures & Algorithms/count-connected-components/submission-9.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:

        adj = {i:[] for i in range(n)}
        for e1,e2 in edges:
            adj[e1].append(e2)
            adj[e2].append(e1)
        vis = set()
        def dfs(node):
            if node in vis:
                return
            vis.add(node)
            for nei in adj[node]:
                dfs(nei)
            return
        
        res = 0
        for i in range(n):
            if i not in vis:
                res+=1
                dfs(i)
        return res

        